File: MPI_Code/Fortran/allgatherv.f90
PROGRAM allgatherv
USE mpi_f08
IMPLICIT NONE
INTEGER::isend(3),irecv(6)
INTEGER::ircnt(0:2), idisp(0:2)
INTEGER::i,iscnt,myrank
ircnt=(/1,2,3/);   idisp=(/0,1,3/)
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,myrank)
DO i=1,myrank+1
  isend(i)=myrank+1
END DO
iscnt=myrank+1
CALL MPI_Allgatherv(isend,iscnt,MPI_INTEGER,irecv,ircnt,idisp,MPI_INTEGER,MPI_COMM_WORLD)
PRINT*,'myrank:',myrank,'irecv=',irecv

irecv=0
IF(myrank==0) irecv(1)=(myrank+1)*10
IF(myrank==1) irecv(2:3)=(myrank+1)*10
IF(myrank==2) irecv(4:6)=(myrank+1)*10
CALL MPI_Allgatherv(MPI_IN_PLACE,iscnt,MPI_INTEGER,irecv,ircnt,idisp,MPI_INTEGER,MPI_COMM_WORLD)
PRINT*,'MYRANK:',myrank,'IRECV=',irecv
CALL MPI_Finalize
END PROGRAM allgatherv

File: example_code/mpi/MPI_Code/C/allgatherv.c
#include <stdio.h>
#include <mpi.h>
int main()
{
  int i,myrank,tmp;
  int isend[3],irecv[6];
  int iscnt,ircnt[3]={1,2,3}, idisp[3]={0,1,3};
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
  for(i=0;i<myrank+1;i++) isend[i]=myrank+1;
  iscnt=myrank+1;
  MPI_Allgatherv(isend,iscnt,MPI_INT,irecv,ircnt,idisp,MPI_INT,MPI_COMM_WORLD);
  printf("myrank: %d    irecv=",myrank); 
  for(i=0;i<6;i++) printf("   %d",irecv[i]);
  printf("\n");

  tmp=(myrank+1)*10;
  if(myrank==0) irecv[0]=tmp;
  if(myrank==1) irecv[1]=irecv[2]=tmp;
  if(myrank==2) irecv[3]=irecv[4]=irecv[5]=tmp;
  MPI_Allgatherv(MPI_IN_PLACE,iscnt,MPI_INT,irecv,ircnt,idisp,MPI_INT,MPI_COMM_WORLD);
  printf("MYRANK: %d    IRECV=",myrank);
  for(i=0;i<6;i++) printf("   %d",irecv[i]);
  printf("\n");


  MPI_Finalize();
  return 0;
}
